File: admin-frontend/src/api/org.ts
import client from './client'
import type { PageResult } from './types'

// ---- 类型定义（对齐后端 schemas/org.py） ----

export interface DepartmentTreeNode {
  id: number
  parent_id: number | null
  name: string
  leader_id: number | null
  sort_order: number
  children: DepartmentTreeNode[]
}

export interface RoleBrief {
  id: number
  role_name: string
  role_code: string
}

export interface RoleOut {
  id: number
  role_name: string
  role_code: string
  description: string | null
}

export interface UserOut {
  id: number
  username: string
  display_name: string | null
  department_id: number | null
  department_name: string | null
  status: number
  roles: RoleBrief[]
}

export interface PermissionItem {
  permission_code: string
  permission_type: string
}

// ---- 部门 ----
export const getDepartments = () =>
  client.get<DepartmentTreeNode[], DepartmentTreeNode[]>('/org/departments')

export const createDepartment = (data: {
  parent_id?: number | null
  name: string
  leader_id?: number | null
  sort_order?: number
}) => client.post<DepartmentTreeNode, DepartmentTreeNode>('/org/departments', data)

export const updateDepartment = (
  id: number,
  data: {
    parent_id?: number | null
    name?: string | null
    leader_id?: number | null
    sort_order?: number | null
  },
) => client.put<DepartmentTreeNode, DepartmentTreeNode>(`/org/departments/${id}`, data)

export const deleteDepartment = (id: number) =>
  client.delete<void, void>(`/org/departments/${id}`)

// ---- 用户 ----
export interface UserListParams {
  keyword?: string
  department_id?: number
  status?: number
  page?: number
  page_size?: number
}

export const getUsers = (params: UserListParams) =>
  client.get<PageResult<UserOut>, PageResult<UserOut>>('/org/users', { params })

export const getUser = (id: number) => client.get<UserOut, UserOut>(`/org/users/${id}`)

export const createUser = (data: {
  username: string
  password: string
  display_name?: string | null
  department_id?: number | null
  role_ids?: number[]
  status?: number
}) => client.post<UserOut, UserOut>('/org/users', data)

export const updateUser = (
  id: number,
  data: { display_name?: string | null; department_id?: number | null; role_ids?: number[] | null },
) => client.put<UserOut, UserOut>(`/org/users/${id}`, data)

export const resetPassword = (id: number, password: string) =>
  client.post<void, void>(`/org/users/${id}/password`, { password })

export const updateUserStatus = (id: number, status: number) =>
  client.patch<void, void>(`/org/users/${id}/status`, { status })

// ---- 角色 ----
export const getRoles = () => client.get<RoleOut[], RoleOut[]>('/org/roles')

export const createRole = (data: {
  role_name: string
  role_code: string
  description?: string | null
}) => client.post<RoleOut, RoleOut>('/org/roles', data)

export const updateRole = (
  id: number,
  data: { role_name?: string | null; description?: string | null },
) => client.put<RoleOut, RoleOut>(`/org/roles/${id}`, data)

export const deleteRole = (id: number) => client.delete<void, void>(`/org/roles/${id}`)

export const getRolePermissions = (id: number) =>
  client.get<PermissionItem[], PermissionItem[]>(`/org/roles/${id}/permissions`)

export const setRolePermissions = (id: number, permissions: PermissionItem[]) =>
  client.post<void, void>(`/org/roles/${id}/permissions`, { permissions })

File: admin/admin/schemas/org.py
"""组织架构相关 Pydantic 模型：部门、用户、角色与权限。"""

from __future__ import annotations

from pydantic import BaseModel, ConfigDict, Field

from admin.schemas.common import PageResult


# ---- 角色 ----
class RoleBrief(BaseModel):
    """角色简要信息（列表/嵌套展示用）。"""

    model_config = ConfigDict(from_attributes=True)

    id: int
    role_name: str
    role_code: str


class RoleCreate(BaseModel):
    role_name: str = Field(..., min_length=1, max_length=64)
    role_code: str = Field(..., min_length=1, max_length=64)
    description: str | None = Field(None, max_length=255)


class RoleUpdate(BaseModel):
    role_name: str | None = Field(None, min_length=1, max_length=64)
    description: str | None = Field(None, max_length=255)


class RoleOut(BaseModel):
    """角色详情输出。"""

    model_config = ConfigDict(from_attributes=True)

    id: int
    role_name: str
    role_code: str
    description: str | None = None


# ---- 角色-操作权限 ----
class PermissionItem(BaseModel):
    permission_code: str = Field(..., min_length=1, max_length=64)
    permission_type: str = Field(..., min_length=1, max_length=16)


class RolePermissionsUpdate(BaseModel):
    """角色权限全量覆盖请求。"""

    permissions: list[PermissionItem] = Field(default_factory=list)


# ---- 部门 ----
class DepartmentCreate(BaseModel):
    parent_id: int | None = None
    name: str = Field(..., min_length=1, max_length=128)
    leader_id: int | None = None
    sort_order: int = 0


class DepartmentUpdate(BaseModel):
    parent_id: int | None = None
    name: str | None = Field(None, min_length=1, max_length=128)
    leader_id: int | None = None
    sort_order: int | None = None


class DepartmentOut(BaseModel):
    """部门节点输出。"""

    model_config = ConfigDict(from_attributes=True)

    id: int
    parent_id: int | None = None
    name: str
    leader_id: int | None = None
    sort_order: int = 0


class DepartmentTreeNode(DepartmentOut):
    """部门树节点（递归携带 children）。"""

    children: list[DepartmentTreeNode] = Field(default_factory=list)


# ---- 用户 ----
class UserCreate(BaseModel):
    username: str = Field(..., min_length=1, max_length=64)
    password: str = Field(..., min_length=6, max_length=128)
    display_name: str | None = Field(None, max_length=128)
    department_id: int | None = None
    role_ids: list[int] = Field(default_factory=list)
    status: int = Field(1, ge=0, le=1)


class UserUpdate(BaseModel):
    display_name: str | None = Field(None, max_length=128)
    department_id: int | None = None
    role_ids: list[int] | None = None


class UserOut(BaseModel):
    """用户列表/详情输出（含部门名与角色）。"""

    id: int
    username: str
    display_name: str | None = None
    department_id: int | None = None
    department_name: str | None = None
    status: int
    roles: list[RoleBrief] = Field(default_factory=list)


class ResetPasswordRequest(BaseModel):
    password: str = Field(..., min_length=6, max_length=128)


class UserStatusUpdate(BaseModel):
    status: int = Field(..., ge=0, le=1)


class UserListResult(PageResult[UserOut]):
    """用户分页响应。"""
